File: api/src/db/indexer.ts
import crypto from 'crypto';
import { Kafka, Consumer, EachMessagePayload } from 'kafkajs';
import pkg from 'pg';
const { Pool } = pkg;

const pool = new Pool({
  connectionString: process.env.DATABASE_URL || 'postgres://obs:obspass@localhost:5432/observability',
  max: 10,
});

interface SessionRow {
  id: string;
  urlSeed: string;
  userAgent: string;
  startedAt: Date;
}

// Maps: sessionId -> { id, urlSeed, userAgent }
const sessionCache = new Map<string, SessionRow>();
const pageCache = new Map<string, string>(); // sessionId -> pageId

const TOPICS = [
  'obs.network.requests',
  'obs.network.responses',
  'obs.dom.mutations',
  'obs.screenshots',
];

export class PostgresIndexer {
  private consumer: Consumer;

  constructor(kafka: Kafka, groupId: string) {
    this.consumer = kafka.consumer({ groupId });
  }

  async start(): Promise<void> {
    await this.consumer.connect();
    for (const t of TOPICS) {
      await this.consumer.subscribe({ topic: t });
    }
    await this.consumer.run({
      eachMessage: async (p: EachMessagePayload) => this.handle(p),
    });
    console.log('[PostgresIndexer] Consumer started');
  }

  private async handle({ topic, message }: EachMessagePayload): Promise<void> {
    try {
      const event = JSON.parse(message.value!.toString());
      switch (topic) {
        case 'obs.network.requests':
          await this.indexRequest(event);
          break;
        case 'obs.network.responses':
          await this.indexResponse(event);
          break;
        case 'obs.dom.mutations':
          await this.indexDomEvent(event);
          break;
        case 'obs.screenshots':
          await this.indexScreenshot(event);
          break;
      }
    } catch (err) {
      console.error('[PostgresIndexer] Error:', err);
    }
  }

  // -----------------------------------------------------------------------
  // Session + page auto-creation
  // -----------------------------------------------------------------------
  private async ensureSession(event: any): Promise<string> {
    const sid = event.sessionId;
    if (!sid) return '';

    if (sessionCache.has(sid)) return sid;

    // Insert session
    const url = event.url ?? event.payload?.url ?? 'unknown';
    const now = new Date(event.ts ?? Date.now());
    try {
      await pool.query(
        `INSERT INTO sessions (id, started_at, url_seed, user_agent)
         VALUES ($1, $2, $3, $4)
         ON CONFLICT (id) DO NOTHING`,
        [sid, now, url, event.payload?.userAgent ?? ''],
      );
    } catch { /* ignore conflict */ }

    sessionCache.set(sid, { id: sid, urlSeed: url, userAgent: '', startedAt: now });
    return sid;
  }

  private async ensurePage(sessionId: string, event: any): Promise<string | null> {
    if (!sessionId) return null;
    if (pageCache.has(sessionId)) return pageCache.get(sessionId)!;

    const url = event.url ?? event.payload?.url ?? 'unknown';
    const now = new Date(event.ts ?? Date.now());

    try {
      const res = await pool.query(
        `INSERT INTO pages (session_id, url, navigated_at, title)
         VALUES ($1, $2, $3, $4)
         ON CONFLICT DO NOTHING
         RETURNING id`,
        [sessionId, url, now, event.payload?.title ?? ''],
      );
      if (res.rows.length > 0) {
        const pid = res.rows[0].id;
        pageCache.set(sessionId, pid);
        return pid;
      }
    } catch { /* ignore */ }

    // Fetch existing
    try {
      const res = await pool.query('SELECT id FROM pages WHERE session_id = $1 LIMIT 1', [sessionId]);
      if (res.rows.length > 0) {
        pageCache.set(sessionId, res.rows[0].id);
        return res.rows[0].id;
      }
    } catch {}
    return null;
  }

  // -----------------------------------------------------------------------
  // Indexers
  // -----------------------------------------------------------------------
  private async indexRequest(event: any): Promise<void> {
    const sid = await this.ensureSession(event);
    const pid = await this.ensurePage(sid, event);

    const id = event.id;
    const ts = new Date(event.ts ?? Date.now());
    const method = event.method ?? 'GET';
    const url = event.url ?? '';
    const urlHost = event.urlHost ?? (() => { try { return new URL(url).hostname; } catch { return ''; } })();
    const resourceType = event.resourceType ?? 'other';
    const headers = event.headers ?? {};

    await pool.query(
      `INSERT INTO requests (id, page_id, ts, method, url, url_host, resource_type, headers)
       VALUES ($1, $2, $3, $4, $5, $6, $7, $8)
       ON CONFLICT (id) DO NOTHING`,
      [id, pid, ts, method, url, urlHost, resourceType, JSON.stringify(headers)],
    );
  }

  private async indexResponse(event: any): Promise<void> {
    // The response 'id' is the normalized UUID that matches the request
    const rid = event.id;
    if (!rid) return;

    const ts = new Date(event.ts ?? Date.now());
    const status = event.status ?? 0;
    const statusText = event.statusText ?? '';
    const headers = event.headers ?? {};
    const bodyRef = event.payload?.body_ref ?? event.body_ref ?? null;
    const bodySize = event.payload?.bodySize ?? null;
    const transferSize = event.transferSize ?? -1;
    const mimeType = event.mimeType ?? '';
    const timing = event.timing ?? null;

    try {
      await pool.query(
        `INSERT INTO responses (request_id, ts, status, status_text, headers, body_ref, body_size, transfer_size, mime_type, timing)
         VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10)`,
        [rid, ts, status, statusText, JSON.stringify(headers), bodyRef, bodySize, transferSize, mimeType, timing ? JSON.stringify(timing) : null],
      );
    } catch (insertErr: any) {
      if (insertErr.code !== '23505') throw insertErr; // 23505 = unique_violation
    }
  }

  private async indexDomEvent(event: any): Promise<void> {
    const sid = await this.ensureSession(event);
    const pid = await this.ensurePage(sid, event);
    if (!pid) return;

    const id = event.id ?? crypto.randomUUID();
    const ts = new Date(event.ts ?? Date.now());
    const tsPageMs = event.tsPage ?? null;
    const mutationType = event.payload?.type ?? event.type ?? 'mutation';
    const targetPath = event.payload?.target ?? null;

    await pool.query(
      `INSERT INTO dom_events (id, page_id, ts, ts_page_ms, mutation_type, target_path, payload)
       VALUES ($1, $2, $3, $4, $5, $6, $7)
       ON CONFLICT (id) DO NOTHING`,
      [id, pid, ts, tsPageMs, mutationType, targetPath, JSON.stringify(event.payload ?? {})],
    );
  }

  private async indexScreenshot(event: any): Promise<void> {
    const sid = event.sessionId;
    if (!sid) return;
    const pid = pageCache.get(sid) ?? null;
    if (!pid) return;

    const id = event.id ?? crypto.randomUUID();
    const ts = new Date(event.ts ?? Date.now());
    const trigger = event.payload?.trigger ?? 'unknown';
    const format = event.payload?.format ?? 'webp';
    const objectKey = event.payload?.objectKey ?? '';
    const perceptualHash = event.payload?.hash ? parseInt(event.payload.hash, 16) : null;

    await pool.query(
      `INSERT INTO screenshots (id, page_id, ts, trigger, format, object_key, perceptual_hash)
       VALUES ($1, $2, $3, $4, $5, $6, $7)
       ON CONFLICT (id) DO NOTHING`,
      [id, pid, ts, trigger, format, objectKey, perceptualHash],
    );
  }

  async disconnect(): Promise<void> {
    await this.consumer.disconnect();
    await pool.end();
  }
}

File: api-rust/src/db/indexer.rs
use chrono::{DateTime, Utc};
use rdkafka::message::{Message, OwnedMessage};
use sqlx::PgPool;
use std::collections::HashMap;
use std::sync::Arc;
use tokio::sync::Mutex;
use uuid::Uuid;

use crate::metrics::KAFKA_MESSAGES_TOTAL;

#[derive(Clone)]
pub struct PostgresIndexer {
    pool: PgPool,
    session_cache: Arc<Mutex<HashMap<String, Uuid>>>,
    page_cache: Arc<Mutex<HashMap<String, Uuid>>>,
}

impl PostgresIndexer {
    pub fn new(pool: PgPool) -> Self {
        Self {
            pool,
            session_cache: Arc::new(Mutex::new(HashMap::new())),
            page_cache: Arc::new(Mutex::new(HashMap::new())),
        }
    }

    pub async fn handle_message(&self, msg: &OwnedMessage) -> anyhow::Result<()> {
        let topic = msg.topic();
        let payload = msg.payload().unwrap_or_default();
        let event: serde_json::Value = serde_json::from_slice(payload)?;

        KAFKA_MESSAGES_TOTAL
            .with_label_values(&[topic, "consumed"])
            .inc();

        match topic {
            "obs.network.requests" => self.index_request(&event).await?,
            "obs.network.responses" => self.index_response(&event).await?,
            "obs.dom.mutations" => self.index_dom_event(&event).await?,
            "obs.screenshots" => self.index_screenshot(&event).await?,
            _ => {}
        }

        Ok(())
    }

    async fn ensure_session(&self, event: &serde_json::Value) -> anyhow::Result<Uuid> {
        let session_id = event
            .get("sessionId")
            .and_then(|v| v.as_str())
            .unwrap_or("")
            .to_string();

        {
            let cache = self.session_cache.lock().await;
            if let Some(id) = cache.get(&session_id) {
                return Ok(*id);
            }
        }

        let id = Uuid::parse_str(&session_id).unwrap_or_else(|_| Uuid::new_v4());
        let started_at: DateTime<Utc> = event
            .get("ts")
            .and_then(|v| v.as_f64())
            .map(|ts| {
                DateTime::from_timestamp_millis(ts as i64).unwrap_or(Utc::now())
            })
            .unwrap_or(Utc::now());

        sqlx::query(
            "INSERT INTO sessions (id, started_at, url_seed) VALUES ($1, $2, $3) ON CONFLICT (id) DO NOTHING"
        )
        .bind(id)
        .bind(started_at)
        .bind(event.get("url").and_then(|v| v.as_str()))
        .execute(&self.pool)
        .await?;

        let mut cache = self.session_cache.lock().await;
        cache.insert(session_id, id);
        Ok(id)
    }

    async fn ensure_page(&self, event: &serde_json::Value, session_id: Uuid) -> anyhow::Result<Uuid> {
        let cache_key = format!("{}:{}", session_id, event.get("pageId").and_then(|v| v.as_str()).unwrap_or(""));

        {
            let cache = self.page_cache.lock().await;
            if let Some(id) = cache.get(&cache_key) {
                return Ok(*id);
            }
        }

        let page_id = event
            .get("pageId")
            .and_then(|v| v.as_str())
            .and_then(|s| Uuid::parse_str(s).ok())
            .unwrap_or_else(Uuid::new_v4);

        let url = event.get("url").and_then(|v| v.as_str()).unwrap_or("unknown");
        let navigated_at: DateTime<Utc> = event
            .get("ts")
            .and_then(|v| v.as_f64())
            .map(|ts| DateTime::from_timestamp_millis(ts as i64).unwrap_or(Utc::now()))
            .unwrap_or(Utc::now());

        let result = sqlx::query(
            "INSERT INTO pages (id, session_id, url, navigated_at) VALUES ($1, $2, $3, $4) ON CONFLICT (id) DO NOTHING RETURNING id"
        )
        .bind(page_id)
        .bind(session_id)
        .bind(url)
        .bind(navigated_at)
        .execute(&self.pool)
        .await?;

        let final_id = if result.rows_affected() > 0 {
            page_id
        } else {
            sqlx::query_scalar::<_, Uuid>("SELECT id FROM pages WHERE id = $1")
                .bind(page_id)
                .fetch_optional(&self.pool)
                .await?
                .unwrap_or(page_id)
        };

        let mut cache = self.page_cache.lock().await;
        cache.insert(cache_key, final_id);
        Ok(final_id)
    }

    async fn index_request(&self, event: &serde_json::Value) -> anyhow::Result<()> {
        let session_id = self.ensure_session(event).await?;
        let page_id = self.ensure_page(event, session_id).await?;

        let id = event
            .get("id")
            .and_then(|v| v.as_str())
            .and_then(|s| Uuid::parse_str(s).ok())
            .unwrap_or_else(Uuid::new_v4);

        let ts = event
            .get("ts")
            .and_then(|v| v.as_f64())
            .map(|ts| DateTime::from_timestamp_millis(ts as i64).unwrap_or(Utc::now()))
            .unwrap_or(Utc::now());

        sqlx::query(
            "INSERT INTO requests (id, page_id, ts, method, url, url_host, resource_type, initiator_type, headers, post_data_ref)
             VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10)
             ON CONFLICT (id) DO NOTHING"
        )
        .bind(id)
        .bind(page_id)
        .bind(ts)
        .bind(event.get("method").and_then(|v| v.as_str()))
        .bind(event.get("url").and_then(|v| v.as_str()).unwrap_or(""))
        .bind(event.get("urlHost").and_then(|v| v.as_str()))
        .bind(event.get("resourceType").and_then(|v| v.as_str()))
        .bind(event.get("initiatorType").and_then(|v| v.as_str()))
        .bind(event.get("headers"))
        .bind(event.get("postData").and_then(|v| v.as_str()))
        .execute(&self.pool)
        .await?;

        Ok(())
    }

    async fn index_response(&self, event: &serde_json::Value) -> anyhow::Result<()> {
        let id = event
            .get("id")
            .and_then(|v| v.as_str())
            .and_then(|s| Uuid::parse_str(s).ok())
            .unwrap_or_else(Uuid::new_v4);

        let ts = event
            .get("ts")
            .and_then(|v| v.as_f64())
            .map(|ts| DateTime::from_timestamp_millis(ts as i64).unwrap_or(Utc::now()))
            .unwrap_or(Utc::now());

        let result = sqlx::query(
            "INSERT INTO responses (request_id, ts, status, status_text, headers, body_ref, body_size, transfer_size, mime_type, timing)
             VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10)"
        )
        .bind(id)
        .bind(ts)
        .bind(event.get("status").and_then(|v| v.as_i64()).map(|s| s as i16))
        .bind(event.get("statusText").and_then(|v| v.as_str()))
        .bind(event.get("headers"))
        .bind(event.get("bodyRef").and_then(|v| v.as_str()))
        .bind(event.get("bodySize").and_then(|v| v.as_i64()).map(|s| s as i32))
        .bind(event.get("transferSize").and_then(|v| v.as_i64()).map(|s| s as i32))
        .bind(event.get("mimeType").and_then(|v| v.as_str()))
        .bind(event.get("timing"))
        .execute(&self.pool)
        .await;

        if let Err(e) = result {
            if let sqlx::Error::Database(ref db) = e {
                if db.code().as_deref() == Some("23505") {
                    return Ok(());
                }
            }
        }

        Ok(())
    }

    async fn index_dom_event(&self, event: &serde_json::Value) -> anyhow::Result<()> {
        let session_id = self.ensure_session(event).await?;
        let page_id = self.ensure_page(event, session_id).await?;

        let id = event
            .get("id")
            .and_then(|v| v.as_str())
            .and_then(|s| Uuid::parse_str(s).ok())
            .unwrap_or_else(Uuid::new_v4);

        let ts = event
            .get("ts")
            .and_then(|v| v.as_f64())
            .map(|ts| DateTime::from_timestamp_millis(ts as i64).unwrap_or(Utc::now()))
            .unwrap_or(Utc::now());

        let ev_type = event.get("type").and_then(|v| v.as_str());

        sqlx::query(
            "INSERT INTO dom_events (id, page_id, ts, ts_page_ms, mutation_type, payload)
             VALUES ($1, $2, $3, $4, $5, $6)"
        )
        .bind(id)
        .bind(page_id)
        .bind(ts)
        .bind(event.get("tsPage").and_then(|v| v.as_f64()))
        .bind(ev_type)
        .bind(event.get("payload"))
        .execute(&self.pool)
        .await?;

        Ok(())
    }

    async fn index_screenshot(&self, event: &serde_json::Value) -> anyhow::Result<()> {
        let session_id_str = event
            .get("sessionId")
            .and_then(|v| v.as_str())
            .unwrap_or("");

        let page_id = {
            let cache = self.page_cache.lock().await;
            cache.get(session_id_str).copied()
        };

        let page_id = match page_id {
            Some(id) => id,
            None => {
                let session_id = self.ensure_session(event).await?;
                self.ensure_page(event, session_id).await?
            }
        };

        let id = event
            .get("id")
            .and_then(|v| v.as_str())
            .and_then(|s| Uuid::parse_str(s).ok())
            .unwrap_or_else(Uuid::new_v4);

        let ts = event
            .get("ts")
            .and_then(|v| v.as_f64())
            .map(|ts| DateTime::from_timestamp_millis(ts as i64).unwrap_or(Utc::now()))
            .unwrap_or(Utc::now());

        let payload = event.get("payload");
        let object_key = payload
            .and_then(|p| p.get("objectKey"))
            .and_then(|v| v.as_str())
            .unwrap_or("");

        let hash_str = payload.and_then(|p| p.get("hash")).and_then(|v| v.as_str());
        let p_hash = hash_str.and_then(|h| i64::from_str_radix(h, 16).ok());

        sqlx::query(
            "INSERT INTO screenshots (id, page_id, ts, trigger, format, width, height, file_size, object_key, perceptual_hash)
             VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10)"
        )
        .bind(id)
        .bind(page_id)
        .bind(ts)
        .bind(payload.and_then(|p| p.get("trigger")).and_then(|v| v.as_str()))
        .bind(payload.and_then(|p| p.get("format")).and_then(|v| v.as_str()))
        .bind(payload.and_then(|p| p.get("width")).and_then(|v| v.as_i64()).map(|w| w as i16))
        .bind(payload.and_then(|p| p.get("height")).and_then(|v| v.as_i64()).map(|h| h as i16))
        .bind(payload.and_then(|p| p.get("fileSize")).and_then(|v| v.as_i64()).map(|s| s as i32))
        .bind(object_key)
        .bind(p_hash)
        .execute(&self.pool)
        .await?;

        Ok(())
    }
}
